micro-rdk/src/esp32/camera.rs

```diff
@@ -1,117 +1,254 @@
+                pin_pclk,
+                xclk_freq_hz,
+                ledc_channel,
+                ledc_timer,
+                pixel_format: PixelFormat::JPEG as u32,
+                frame_size,
+                jpeg_quality,
+                // Number of frame buffers to be allocated.
+                // If more than one, then each frame will be acquired (double speed)
+                // when pixel_format == jpeg, fb_count > 1 goes to continuous mode, may need to adjust
+                // xclk_freq_hz down to 10_000_000.
                 fb_count: 1,
                 grab_mode: 0,
                 fb_location: 0,
-                sccb_i2c_port: 0,
+                sccb_i2c_port,
             },
-            last_grab: t.now(),
-        }
+        };
+
+        esp!(unsafe { esp_camera_init(&cam.config) })
+            .map_err(|e| CameraError::InitError(Box::new(e)))?;
+        Ok(Arc::new(Mutex::new(cam)))
     }
-    pub fn setup(&self) -> Result<(), CameraError> {
-        let ret = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_init(&self.config) })
-            as crate::esp32::esp_idf_svc::sys::esp_err_t;
-        let ret = crate::esp32::esp_idf_svc::sys::EspError::convert(ret);
-        ret.map_err(|e| CameraError::CameraInitError(e.into()))
+
+    fn get_frame(&mut self) -> Result<Esp32CameraFrameBuffer, CameraError> {
+        Esp32CameraFrameBuffer::get().ok_or(CameraError::FailedToGetImage)
     }
-    pub fn get_cam_frame(&self) -> Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t> {
-        let ptr = (unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_get() })
-            as *mut crate::esp32::esp_idf_svc::sys::camera_fb_t;
-        if ptr.is_null() {
-            None
-        } else {
-            Some(ptr)
+}
+
+impl Camera for Esp32Camera {
+    fn get_image(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        let frame = self.get_frame()?;
+        if frame.len() > buffer.capacity() {
+            return Err(CameraError::ImageTooBig);
         }
+        let msg = camera::v1::GetImageResponse {
+            mime_type: "image/jpeg".to_string(),
+            image: frame.as_bytes(),
+        };
+        // safety: message must be encoded before the frame is dropped from scope
+        msg.encode(&mut buffer).unwrap();
+
+        return Ok(buffer);
     }
-    pub fn return_cam_frame(
-        &self,
-        frame: Option<*mut crate::esp32::esp_idf_svc::sys::camera_fb_t>,
-    ) {
-        if let Some(ptr) = frame {
-            unsafe { crate::esp32::esp_idf_svc::sys::esp_camera_fb_return(ptr) }
+
+    fn render_frame(&mut self, mut buffer: BytesMut) -> Result<BytesMut, CameraError> {
+        let frame = self.get_frame()?;
+        if frame.len() > buffer.capacity() {
+            return Err(CameraError::ImageTooBig);
         }
+        let msg = HttpBody {
+            content_type: "image/jpeg".to_string(),
+            data: frame.as_bytes().to_vec(),
```

Comment:
does this `to_vec` call make another copy? Is it worth having a separate `as_vec` function on `Esp32CameraFrameBuffer` that calls `Vec::from_raw_parts` directly instead of going through `Bytes`?

Reply:
I added the `Vec::from_raw_parts` but I'm concerned about the safety of it.

As it's now a `Vec`, rust will try to deallocate the data on drop, [but there doesn't appear to be deallocation, just setting bitflags](https://github.com/espressif/esp32-camera/blob/7aa37d4f22503fdac9ccd449e4678c4894c40055/driver/cam_hal.c#L522).